### PR TITLE
Serializable collections

### PR DESCRIPTION
**Pull request by lorenzo**
I dreamt again of the solution to this problem, and it turns out it works :D

Fixes https://github.com/cakephp/cakephp/issues/6910

### CODE REVIEW

File: tests/TestCase/Collection/CollectionTest.php [hunk 1]
```diff
@@ -1433,4 +1433,69 @@ public function testUnfoldedExtract()
...
+        $collection = new Collection([1, 2, 3]);
+        $collection = $collection->append(['a' => 4, 'b' => 5, 'c' => 6]);
+        $selialized = serialize($collection);
+        $unserialized = unserialize($selialized);
+        $this->assertEquals($collection->toList(), $unserialized->toList());
+        $this->assertEquals($collection->toArray(), $unserialized->toArray());
+    }
+
+    /**
+     * Tests serialization when using append
```
Comment (dereuromark) [inline tests/TestCase/Collection/CollectionTest.php]: copy and paste

**Comment by markstory** [comment]
Awesome work @lorenzo! This is much simpler than the previous attempt.

File: tests/TestCase/Collection/CollectionTest.php (hunk 1 above)
Comment (lorenzo) [inline tests/TestCase/Collection/CollectionTest.php]: Fixed

**Comment by lorenzo** [comment]
@markstory I'm happy it turned out to be a simple fix at the end

**Comment by ADmad** [comment]
@lorenzo Dream on sir, dream on!!

File: src/Collection/Collection.php [hunk 2]
```diff
@@ -50,6 +51,28 @@ public function __construct($items)
...
+        return serialize($this->buffered());
+    }
+
+    /**
+     * Unserializes the passed string and rebuilds the Collection instance
+     *
+     * @param string $collection The serialized collection
+     * @return void
+     */
+    public function unserialize($collection)
```
Comment (ADmad) [inline src/Collection/Collection.php]: How about naming the param `$string` instead?

Comment (lorenzo) [inline src/Collection/Collection.php]: I prefer what we have now since it tells me what is in the string

Comment (ADmad) [inline src/Collection/Collection.php]: Without reading the docblock my first intuition is `$collection` suggests it a collection not string. But okay.

**Pull request merged**